### PR TITLE
[SPARK-37713][K8S] Assign namespace to executor configmap

### PR DESCRIPTION
**Pull request by dcoliversun**
### What changes were proposed in this pull request?

Since Spark 3.X, Executor pod needs separate executor configmap. But, no namespace is assigned in configmap while building it. K8s views configmap without namespace as global resource. Once pod access is restricted (global resources cannot be read), and executor cannot obtain configmap itself.

### Why are the changes needed?

Fix executor pod authorization error in K8S. When executor pod cannot read global resources, executor configmap is not available for executor.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass Integration Test and CIs.

### CODE REVIEW

**Comment by dcoliversun** [comment]
@dongjoon-hyun  Would you have review for these code? Thanks :)

**Review by dongjoon-hyun — COMMENTED** [review]
Could you add a test case, @dcoliversun ?

**Comment by dongjoon-hyun** [comment]
Gentle ping, @dcoliversun .

**Comment by dcoliversun** [comment]
> 

Sorry so late to reply. I've been working on it for days. I couldn't add test case because k8s allows configmap without .metadata.namespce. Request to APIServer will put namespace to body. But, .metadata.namespace is still null when request to APIServer. I think it's a good action. User will see this configmap through dashboard. The action that namespace of executor configmap yaml is null will confuse user.

**Review by dongjoon-hyun — COMMENTED** [review]
Why is it a reason not to add a test case? You need to add a test case for your code, not K8s code itself.
> I couldn't add test case because k8s allows configmap without .metadata.namespce.

It sounds like you did try to make K8s itself fail, but we know that Apache Spark works well with the previous config without any issue.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala [hunk 1]
```diff
@@ -87,11 +87,15 @@ private[spark] object KubernetesClientUtils extends Logging {
    * Build a Config Map that will hold the content for environment variable SPARK_CONF_DIR
    * on remote pods.
    */
-  def buildConfigMap(configMapName: String, confFileMap: Map[String, String],
-      withLabels: Map[String, String] = Map()): ConfigMap = {
+  def buildConfigMap(
+     configMapName: String,
+     confFileMap: Map[String, String],
+     configMapNameSpace: String,
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala:93]: Please try to get namespace via `confFileMap` instead changing this method signature.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I made two requests for this PR.
- Provide a test coverage for newly added code path.
- Avoid a method signature change as much as possible when you can do.

For the other code, it looks reasonable, @dcoliversun .

**Comment by dcoliversun** [comment]
> 
OK，I understand. I am working on it

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala [hunk 2]
```diff
@@ -76,4 +80,26 @@ class KubernetesClientUtilsSuite extends SparkFunSuite with BeforeAndAfter {
       "testConf.3" -> "test123456")
     assert(output === expectedOutput)
   }
+
+  test("verify that configmap built as expected") {
+    val configMapName = "testConfigMapName"
+    val properties = Map(Config.KUBERNETES_NAMESPACE.key -> "default")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala:86]: If you don't mind, could you use a random name here? Although the previous code doesn't attache the namespace, this `default` is too generic to be in the test case.

Comment (dcoliversun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala:86]: OK, I change it.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala (hunk 1 above)
Comment (dcoliversun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala]: Done, buildConfigMap method could get namespace via `confFileMap`

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala [hunk 3]
```diff
@@ -82,8 +81,9 @@ class KubernetesClientUtilsSuite extends SparkFunSuite with BeforeAndAfter {
   }
 
   test("verify that configmap built as expected") {
-    val configMapName = "testConfigMapName"
-    val properties = Map(Config.KUBERNETES_NAMESPACE.key -> "default")
+    val configMapName = java.util.UUID.randomUUID.toString
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala:84]: `import java.util.UUID`?

Comment (dcoliversun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala:84]: It works without "import java.util.UUID". Is this a code import style limitation? If true, I can import :)

**Comment by dcoliversun** [comment]
Sorry, CI has some error. I'm fixing it.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala:84]: @dcoliversun . Apache Spark recommends not to hide your import in the code. Especially, you are using twice at line 84 and 85. Your code is verbose due to the repeatability and not good at the readability too.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala (hunk 3 above)
Comment (dcoliversun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtilsSuite.scala]: @dongjoon-hyun Thanks for your suggestion. I import java.util.UUID and add name/namespace prefix to improve readability. Please reivew it 😊

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @dcoliversun .

**Comment by dongjoon-hyun** [comment]
cc @williamhyun

**Comment by dcoliversun** [comment]
> +1, LGTM. Thank you, @dcoliversun .

@dongjoon-hyun Thanks for your review, which helps me a lot. HAPPY NEW YEAR

**Comment by dongjoon-hyun** [comment]
Happy New Year, @dcoliversun ! :)

**Comment by dongjoon-hyun** [comment]
K8s tests passed. Merged to master.

**Comment by dongjoon-hyun** [comment]
If you need this in next Apache Spark 3.2.1, please make a backporting PR against `branch-3.2`.

**Comment by Yikun** [comment]
FYI, this PR breaks the basic case to use ConfigMap with namespace specified (driver side), see https://github.com/apache/spark/pull/35215 .

**Comment by Yikun** [comment]
```scala
val builder2 = new ConfigMapBuilder()
      .withNewMetadata()
        // We don't specify ns2 in here
        .withName("configmap2")
      .endMetadata()
      .addToData("testkey", "testvalue")
      .build()
    val configMap2 = client.inNamespace("ns2").resource(builder2).createOrReplace()
    assert(configMap2.getMetadata.getNamespace === "ns2")
```

I just wanna sure, do we really need this PR?

The case shows even if we don't set namespace in configmap, will just set to client default namespace which set in [createKubernetesClient](https://github.com/apache/spark/blob/ddc77fb906cb3ce1567d277c2d0850104c89ac25/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala#L105) with [conf namespace](https://github.com/apache/spark/blob/ddc77fb906cb3ce1567d277c2d0850104c89ac25/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala#L80).

That is even without this PR, we can also set configmap namespace as expected, or did I miss something?

**Comment by GabeChurch** [comment]
Noticing the same behavior as Yikun with this pull breaking ConfigMap usage in non "default" namespace(s).

**Pull request closed without merging**